tree/tree/src/TTreeCache.cxx

```diff
@@ -614,10 +586,296 @@ Int_t TTreeCache::DropBranch(const char *bname, Bool_t subbranches /*= kFALSE*/)
+///
+/// Returns true if we were able to pull the data into the miss cache.
+///
+Bool_t TTreeCache::ProcessMiss(Long64_t pos, int len)
+{
+
+   Bool_t firstMiss = kFALSE;
+   if (fFirstMiss == -1) {
+      fFirstMiss = fEntryCurrent;
+      firstMiss = kTRUE;
+   }
+   fLastMiss = fEntryCurrent;
+   // The first time this is executed, we try to pull in as much data as we can.
+   TBranch *b = CalculateMissEntries(pos, len, firstMiss);
+   if (!b) {
+      if (!firstMiss) {
+         // TODO: this recalculates for *all* branches, throwing away the above work.
+         b = CalculateMissEntries(pos, len, kTRUE);
+      }
+      if (!b) {
+         // printf("ProcessMiss: pos %ld does not appear to correspond to a buffer in this file.\n", pos);
+         // We have gone through all the branches in this file and the requested basket
+         // doesn't appear to be in any of them.  Likely a logic error / bug.
+         fMissCache->fEntries.clear();
+         return kFALSE;
+      }
+   }
+   // TODO: this should be a set.
+   fMissCache->fBranches.push_back(b);
+
+   // OK, sort the entries
+   std::sort(fMissCache->fEntries.begin(), fMissCache->fEntries.end());
+
+   // Now, fetch the buffer.
+   std::vector<Long64_t> positions;
+   positions.reserve(fMissCache->fEntries.size());
+   std::vector<Int_t> lengths;
+   lengths.reserve(fMissCache->fEntries.size());
+   ULong64_t cumulative = 0;
+   for (auto &mcentry : fMissCache->fEntries) {
+      positions.push_back(mcentry.fIO.fPos);
+      lengths.push_back(mcentry.fIO.fLen);
+      mcentry.fIndex = cumulative;
+      cumulative += mcentry.fIO.fLen;
+   }
+   fMissCache->fData.reserve(cumulative);
+   // printf("Reading %lu bytes into miss cache for %lu entries.\n", cumulative, fEntries->size());
+   fNMissReadPref += fMissCache->fEntries.size();
+   fFile->ReadBuffers(&(fMissCache->fData[0]), &(positions[0]), &(lengths[0]), fMissCache->fEntries.size());
+   fFirstMiss = fLastMiss = fEntryCurrent;
+
+   return kTRUE;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Given an IO operation (pos, len) that was a cache miss in the primary TTC,
+/// try the operation again with the miss cache.
+///
+/// Returns true if the IO operation was successful and the contents of buf
+/// were populated with the requested data.
+///
+Bool_t TTreeCache::CheckMissCache(char *buf, Long64_t pos, int len)
+{
+
+   if (!fOptimizeMisses) {
+      return kFALSE;
+   }
+   if (R__unlikely((pos < 0) || (len < 0))) {
+      return kFALSE;
+   }
+
+   // printf("Checking the miss cache for offset=%ld, length=%d\n", pos, len);
+
+   // First, binary search to see if the desired basket is already cached.
+   MissCache::Entry mcentry{IOPos{pos, len}};
+   auto iter = std::lower_bound(fMissCache->fEntries.begin(), fMissCache->fEntries.end(), mcentry);
+
+   if (iter != fMissCache->fEntries.end()) {
+      if (len > iter->fIO.fLen) {
+         fNMissReadMiss++;
```

Comment:
Could you use prefix?

Reply:
Hi @pcanal -

Running through open PRs.  The above item is the only requested change, but (perhaps it's just the fact it is late) I can't quite figure out what your comment above means.  Could you elaborate a bit?

Brian

Reply:
I mean 'prefix addition'
```
++fNMissReadMiss;
```
Not an essential change.